### ISSUE TITLE
Copying source code to runs directory?

### ISSUE THREAD

**Issue by solarflarefx** [guildai/guildai#266]
Perhaps I am misunderstanding, but I am running a simple training script to test feasibility.  I am able to save my tensorboard logs and checkpoints, but it seems that the source code is not being copied to the runs directory.  Is this supposed to be specified?  

I am running in Windows using an anaconda environment.  My runs save to: ANACONDA_ENV_PATH\.guild\runs\RUN_NUM

I get three directorys:
.guild
ckpts
logs

ckpts are where I saved my checkpoints, logs are where tensorboard saves its params and .guild seems to have 'opref' and 'output' files and and 'attrs' dir, which has the following files to it:
![image](https://user-images.githubusercontent.com/45402633/107894370-69752c00-6efd-11eb-91a6-99b7cfa94605.png)

Eventually I would like to run this on a full data pipeline that consists of many python files in a directory.  Is it possible to copy over all these scripts (source code) to the run directory?

**Comment by gar1t** [comment]
Source code for each run is stored under `.guild/sourcecode` by default. You can list the source code files by using the `--sourcecode` option with `ls`.

    guild ls --sourcecode

You can open a source code directory the same way with the `open` command.

    guild open --sourcecode

Guild stores source code in a directory under `.guild` so it doesn't conflict with required and generated files. If you want to save source code in the run directory root, you can config the operation `sourcecode` attribute this way:

``` yaml
op:
  sourcecode:
    dest: .
```

For more info see [Guild File Reference](https://my.guild.ai/t/guild-file-reference/197#sourcecode-dest).

**Comment by solarflarefx** [comment]
@gar1t Thanks for the reply.  So I created a guild.yml file as follows:

``` yaml
train:
  description: Train a model using a Python script
  main: train
op:
  sourcecode:
    dest: .
```

However, the sourcecode does not appear to go into the run directory.  I tried running the --sourcecode commands you suggested but the file could not be found.  

I am simply running a basic tensorboard tutorial script that I named as train.py:

	import tensorflow as tf
	import datetime
	import os

	def create_model():
	  return tf.keras.models.Sequential([
	    tf.keras.layers.Flatten(input_shape=(28, 28)),
	    tf.keras.layers.Dense(512, activation='relu'),
	    tf.keras.layers.Dropout(0.2),
	    tf.keras.layers.Dense(10, activation='softmax')
	  ])



	def main():

	  def train_step(model, optimizer, x_train, y_train):
	    with tf.GradientTape() as tape:
	      predictions = model(x_train, training=True)
	      loss = loss_object(y_train, predictions)
	    grads = tape.gradient(loss, model.trainable_variables)
	    optimizer.apply_gradients(zip(grads, model.trainable_variables))

	    train_loss(loss)
	    train_accuracy(y_train, predictions)

	  def test_step(model, x_test, y_test):
	    predictions = model(x_test)
	    loss = loss_object(y_test, predictions)

	    test_loss(loss)
	    test_accuracy(y_test, predictions)

	  mnist = tf.keras.datasets.mnist

	  (x_train, y_train), (x_test, y_test) = mnist.load_data()
	  x_train, x_test = x_train / 255.0, x_test / 255.0

	  train_dataset = tf.data.Dataset.from_tensor_slices((x_train, y_train))
	  test_dataset = tf.data.Dataset.from_tensor_slices((x_test, y_test))

	  train_dataset = train_dataset.shuffle(60000).batch(64)
	  test_dataset = test_dataset.batch(64)

	  loss_object = tf.keras.losses.SparseCategoricalCrossentropy()
	  optimizer = tf.keras.optimizers.Adam()

	  # Define our metrics
	  train_loss = tf.keras.metrics.Mean('train_loss', dtype=tf.float32)
	  train_accuracy = tf.keras.metrics.SparseCategoricalAccuracy('train_accuracy')
	  test_loss = tf.keras.metrics.Mean('test_loss', dtype=tf.float32)
	  test_accuracy = tf.keras.metrics.SparseCategoricalAccuracy('test_accuracy')

	  current_time = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
	  train_log_dir = 'logs/gradient_tape/' + current_time + '/train'
	  test_log_dir = 'logs/gradient_tape/' + current_time + '/test'
	  train_summary_writer = tf.summary.create_file_writer(train_log_dir)
	  test_summary_writer = tf.summary.create_file_writer(test_log_dir)

	  model = create_model() # reset our model
	  checkpoint_directory = "ckpts"
	  checkpoint_prefix = os.path.join(checkpoint_directory, "ckpt")
	  checkpoint = tf.train.Checkpoint(optimizer=optimizer, model=model)

	  EPOCHS = 5

	  for epoch in range(EPOCHS):
	    for (x_train, y_train) in train_dataset:
	      train_step(model, optimizer, x_train, y_train)
	    with train_summary_writer.as_default():
	      tf.summary.scalar('loss', train_loss.result(), step=epoch)
	      tf.summary.scalar('accuracy', train_accuracy.result(), step=epoch)

	    for (x_test, y_test) in test_dataset:
	      test_step(model, x_test, y_test)
	    with test_summary_writer.as_default():
	      tf.summary.scalar('loss', test_loss.result(), step=epoch)
	      tf.summary.scalar('accuracy', test_accuracy.result(), step=epoch)

	    template = 'Epoch {}, Loss: {}, Accuracy: {}, Test Loss: {}, Test Accuracy: {}'
	    print (template.format(epoch+1,
		                   train_loss.result(),
		                   train_accuracy.result()*100,
		                   test_loss.result(),
		                   test_accuracy.result()*100))

	    checkpoint.save(file_prefix=checkpoint_prefix)

	    # Reset metrics every epoch
	    train_loss.reset_states()
	    test_loss.reset_states()
	    train_accuracy.reset_states()
	    test_accuracy.reset_states()

	if __name__ == "__main__":
	    from guild import ipy as guild
	    guild.run(main)

Am I misunderstanding something here?  Everything seems to be copying to the run directory except for the source script train.py.

**Comment by solarflarefx** [comment]
OK it seems that the issue is related to running the code from PyCharm.  I followed the guideline here: https://my.guild.ai/t/debugging-guild-operations-in-pycharm/453

If I simply run the debugger then the sourcecode does not seem to copy anywhere.

If instead I run from bash: guild run train

This does seem to save the sourcecode.  However, it seems that the sourcecode saves in a separate run directory.

1. Is there some additional step that must be taken to run the yaml file?
2. Did I make a mistake in my guild.yml file if I wanted the sourcecode to be in the same directory as the ckpts and logs?

**Comment by gar1t** [comment]
I took the liberty of formatting your YAML example so I could see the indent levels.

It should look like this:

``` yaml
train:
  description: Train a model using a Python script
  main: train
  sourcecode:
    dest: .
```

You can omit `main: train` there because the operation is named train and uses that module by default.

The `op` that I used was just an example. In your case the operation is `train` so that's where the `sourcecode` attribute goes.

You can test what's copied for a run using the `--test-sourcecode` option.

    $ guild run train --test-sourcecode

I'll look over that PyCharm guide. Something may not be working as expected.
